### PR TITLE
document and adopt best practices for using `.ts` for typedefs

### PR DESCRIPTION
**Pull request by turadg**
refs: #10086

## Description

Document the practice established in https://github.com/Agoric/agoric-sdk/pull/10149 into a Best Practice.

Adopt it in most packages. I only left these `types.d.ts`
-  swing-store because the file is only re-exports, nothing to typecheck
- smart-wallet because there's no entrypoint for the package

### Security Considerations
none

### Scaling Considerations
none

### Documentation Considerations
per se

### Testing Considerations
https://github.com/Agoric/agoric-sdk/pull/10149#issuecomment-2372759016 asked for more testing. This PR was originally intended to close the issue but the regression test is complicated enough and this PR has gotten big enough that I think it should land on its own.

### Upgrade Considerations
n/a

### CODE REVIEW

**Comment by mhofman** [comment]
> @mhofman is there something more than [5952bbd](https://github.com/Agoric/agoric-sdk/commit/5952bbd0ae907f442b7ed5de24d44ea5a41349f9) you're after?

I think a test snapshot is a little heavy-handed in this case as any change in the shape of exports (deep apparently) causes the snapshot content to change. We're technically interested that:
- Runtime import of `.index.js` does not fail. I believe this can be either a static or dynamic import.
- As mentioned by @0xpatrickdev, we should verify the types are available through `@agoric/orchestration` type imports. I think a TS test file importing using that specifier is sufficient. I don't recall if that needs to be tested from a dependent package, or if same package is sufficient. At runtime in Node.js, a same package reference is fine, and [`@endo/init` actually uses that trick](https://github.com/endojs/endo/blob/392e9b217247d220bb8d6d3d906eb233b91a95c0/packages/init/pre-bundle-source.js#L7-L9), but maybe TSC has a different opinion (eslint surely complains).

File: docs/typescript.md
Comment (mhofman) [inline docs/typescript.md:45]: I would add a sentence clarifying what's going on:

> The actual type implementation is then written in `types.ts` and `other-types.ts` files. These files are never runtime imported as they are only linked through a `.d.ts` file.

**Review by mhofman — COMMENTED** [review]
Overall I like the direction! Thanks for taking this on.

> we can try getting them all int his PR if you like

Your call, I'm fine with trying if it's not too much work.

**Comment by turadg** [comment]
I think to have a real regression test for https://github.com/Agoric/agoric-sdk/issues/10086 we'll need to check exports after the package is packed.

I'm going to leave that out of scope of this PR so I've removed the "closes".

File: tsconfig-build-options.json
Comment (mhofman) [inline tsconfig-build-options.json:9]: Oh comments are supported in these `.json` files!

Looks like for [local resolution, `.jsonc` seem to be supported](https://redirect.github.com/microsoft/TypeScript/issues/43121). Wondering if we should switch to that?

File: tsconfig-build-options.json
Comment (mhofman) [inline tsconfig-build-options.json:10]: I was first surprised this is necessary, but I guess when I used to manually author `.d.ts` files, I either was not emitting from `.js`, or I kept any `.js` out of reach of tsc. Thinking about it, there is no way for tsc to know that the `.d.ts` file was manually authored and not previously generated.

File: packages/notifier/src/types.ts
Comment (mhofman) [inline packages/notifier/src/types.ts:1]: I have not reviewed that this file is equivalent to the former `types.js` file. Was this purely generated by tsc or were any edits manually made?

**Review by mhofman — APPROVED** [review]
> I think to have a real regression test for https://github.com/Agoric/agoric-sdk/issues/10086 we'll need to check exports after the package is packed.

Technically for #10086, we originally just cared that a runtime import of `@agoric/orchestration` did not fail. Now in #10149 we somewhat decided to expend the scope of that issue to also verify that type imports through `@agoric/orchestration` also work.

In a perfect world, I agree we should verify that through the packed NPM package. However I think it's sufficient for now to verify that in a `.ts` test file, that for example `import type { Orchestrator } from '@agoric/orchestration';` doesn't fail. I believe this would fail if for example the `types-index.d.ts` file forgot to include the correct `export *`, or if the `index.js` failed to `export * from './src/types-index.js'`.

If you prefer to postpone adding a check like that (or if it doesn't work as I think), feel free to ignore in this PR.

> 10086 typecheck typedefs
 #10158

Can we rename this PR before merging since mergify uses the PR title in the merge commit?

File: packages/notifier/src/types.ts
Comment (turadg) [reply packages/notifier/src/types.ts:1]: purely by TSC. maybe plus an import that had been missing

File: tsconfig-build-options.json
Comment (turadg) [reply tsconfig-build-options.json:9]: Good question but I don't think it's worth it because we couldn't do it for all the tsconfig

**Pull request merged**
